optimistix/_solver/bisection.py

```diff
@@ -18,6 +20,61 @@ class _BisectionState(eqx.Module, strict=True):
     error: Float[Array, ""]
 
 
+class _ExpansionCarry(eqx.Module, strict=True):
+    lower: Scalar
+    upper: Scalar
+
+
+def _interval_contains_root(carry: _ExpansionCarry,
+                            /,
+                            *,
+                            need_positive: Bool[Array, ""],
+                            expand_upper: Bool[Array, ""],
+                            fn: Fn[Scalar, Scalar, Aux],
+                            args: Any,
+                            ) -> Bool[Array, ""]:
+    new_boundary = jnp.where(expand_upper, carry.upper, carry.lower)
+    carry_val, _ = fn(new_boundary, args)
+    return need_positive ^ (carry_val > 0.0)
+
+
+def _expand_interval(carry: _ExpansionCarry,
+                     /,
+                     *,
+                     expand_upper: Bool[Array, ""],
+                     ) -> _ExpansionCarry:
+    new_domain = 2.0 * (carry.upper - carry.lower)
+    new_lower = jnp.where(expand_upper, carry.upper, carry.lower - new_domain)
+    new_upper = jnp.where(expand_upper, carry.upper + new_domain, carry.lower)
+    return _ExpansionCarry(new_lower, new_upper)
+
+
+def _expand_interval_repeatedly(initial_interval: _ExpansionCarry,
+                                upper_val: Scalar,
+                                lower_val: Scalar,
+                                need_positive: Bool[Array, ""],
+                                *,
+                                fn: Fn[Scalar, Scalar, Any],
+                                args: PyTree,
+                                ) -> _ExpansionCarry:
+    expand_upper = need_positive ^ (upper_val < lower_val)
+    cond_fun = ft.partial(_interval_contains_root,
+                          need_positive=need_positive,
+                          expand_upper=expand_upper,
+                          fn=fn,
+                          args=args)
+    body_fun = ft.partial(_expand_interval, expand_upper=expand_upper)
+    return eqxi.while_loop(cond_fun, body_fun, initial_interval, kind="lax")
+
+
+def _retain_interval(initial_interval: _ExpansionCarry,
+                     upper_val: Scalar,
+                     lower_val: Scalar,
+                     need_positive: Bool[Array, ""],
```

Comment:
Nit: I think we can simplify the `cond` by partially-applying these arguments to the `_expand_interval_repeatedly` branch, so that only the `_ExpansionCarry` object is passed to `lax.cond(..., initial_interval)`.

Reply:
Good tip for the future, but now that I've taken your other suggestion, this isn't relevant anymore.